clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java

```diff
@@ -0,0 +1,121 @@
+ */
+public class RetryManager implements Closeable {
+  private static final Logger LOGGER = LogManager.getLogger(RetryManager.class);
+  private static final int TOKEN_BUCKET_REFILL_INTERVAL_IN_SECONDS = 1;
+  private static final int TOKEN_BUCKET_CAPACITY_MULTIPLE = 5;
+  private final AtomicBoolean retryBudgetEnabled = new AtomicBoolean();
+  private final long enforcementWindowInMs;
+  private final double retryBudgetInPercentDecimal;
+  private final AtomicLong requestCount = new AtomicLong();
+  private final AtomicReference<TokenBucket> retryTokenBucket = new AtomicReference<>(null);
+  private final ScheduledExecutorService scheduler;
+  private final Clock clock;
+
+  private long lastUpdateTimestamp;
+  private long previousQPS = 0;
+
+  public RetryManager(long enforcementWindowInMs, double retryBudgetInPercentDecimal, Clock clock) {
+    if (enforcementWindowInMs <= 0 || retryBudgetInPercentDecimal <= 0) {
+      scheduler = null;
+      retryBudgetEnabled.set(false);
+    } else {
+      scheduler = Executors.newScheduledThreadPool(1);
+      retryBudgetEnabled.set(true);
+      lastUpdateTimestamp = clock.millis();
+      scheduler.schedule(this::updateRetryTokenBucket, enforcementWindowInMs, TimeUnit.MILLISECONDS);
+    }
+    this.enforcementWindowInMs = enforcementWindowInMs;
+    this.retryBudgetInPercentDecimal = retryBudgetInPercentDecimal;
+    this.clock = clock;
+  }
+
+  public RetryManager(long enforcementWindowInMs, double retryBudgetInPercentDecimal) {
+    this(enforcementWindowInMs, retryBudgetInPercentDecimal, Clock.systemUTC());
+  }
+
+  public void recordRequest() {
+    if (retryBudgetEnabled.get()) {
+      requestCount.incrementAndGet();
+    }
+  }
+
+  public boolean isRetryAllowed() {
+    if (!retryBudgetEnabled.get() || retryTokenBucket.get() == null) {
+      // All retries are allowed when the feature is disabled or during the very first enforcement window when we
+      // haven't collected enough data points yet
+      return true;
+    }
+    return retryTokenBucket.get().tryConsume(1);
+  }
+
+  private void updateRetryTokenBucket() {
+    if (retryBudgetEnabled.get() && requestCount.get() > 0) {
+      try {
+        long elapsedTimeInMs = clock.millis() - lastUpdateTimestamp;
+        long requestCountSinceLastUpdate = requestCount.getAndSet(0);
+        lastUpdateTimestamp = clock.millis();
+        // Minimum user request per second will be 1
+        long newQPS = (long) Math
+            .ceil((double) requestCountSinceLastUpdate / (double) TimeUnit.MILLISECONDS.toSeconds(elapsedTimeInMs));
+        if (previousQPS > 0) {
+          long difference = Math.abs(previousQPS - newQPS);
+          double differenceInPercentDecimal = (double) difference / (double) previousQPS;
+          if (differenceInPercentDecimal > 0.1) {
+            // Only update the retry token bucket if the change in request per seconds is more than 10 percent
```

Comment:
Why can't we update the token bucket all the time?
Is it an expensive operation?

Reply:
It's not an expensive operation. However, the default initial value for token bucket is `capacity` which is maxed out. If we keep initializing new token bucket then we are technically allowing more retry even if the traffic remains the same.

Reply:
Emm.. 
So the initialization would assume the max cap, and I thought it should be initialized by the 0.1 * previous QPS, no? Maybe not for the first time, but for the subsequent initialization, it doesn't necessary to initialize with max cap, right?

Reply:
I meant the max cap as in the 0.1 * previous QPS * `TOKEN_BUCKET_CAPACITY_MULTIPLE` (5).

Reply:
What is the reason to multiple `TOKEN_BUCKET_CAPACITY_MULTIPLE`? To tolerate some random QPS spike?

Reply:
Does this mean fast-client will allow 50% of retry all the time or just some short period and the bursty retry will still draw from the total budget, right?